### PR TITLE
Improve tower client caching

### PR DESCRIPTION
**Pull request by pditommaso**
This PR implements a more aggressive caching for HTTP request submitted to Platform assuming credentials cannot be changed during a workflow execution, and therefore caching all requests having the same workflow Id

### CODE REVIEW

**Comment by pditommaso** [comment]
@munishchouhan can you please review and test this? (I'm preparing a test release in stage)

**Comment by munishchouhan** [comment]
> @munishchouhan can you please review and test this? (I'm preparing a test release in stage)

ok sure

**Comment by pditommaso** [comment]
I'm getting this, four identical requests mostly at the same time. However it should not happen  

```
2024-12-16 10:56:20.828	09:56:20.826 [virtual-executor116] TRACE i.s.w.t.c.connector.TowerConnector - Tower GET 'https://api.cloud.stage-seqera.io/user-info' - can refresh=true; msgId=170f93d451ab; tokens=JwtAuth{key='fa657055338a584c9483fd732bb0da01', endpoint='https://api.cloud.stage-seqera.io', bearer='eyJ0aWQiOiAyNTV9LmYyNjBjO..', refresh='null', createdAt=null, updatedAt=null}
2024-12-16 10:56:20.828	09:56:20.826 [virtual-executor117] TRACE i.s.w.t.c.connector.TowerConnector - Tower GET 'https://api.cloud.stage-seqera.io/user-info' - can refresh=true; msgId=5d73e356292f; tokens=JwtAuth{key='fa657055338a584c9483fd732bb0da01', endpoint='https://api.cloud.stage-seqera.io', bearer='eyJ0aWQiOiAyNTV9LmYyNjBjO..', refresh='null', createdAt=null, updatedAt=null}
2024-12-16 10:56:20.828	09:56:20.827 [virtual-executor114] TRACE i.s.w.t.c.connector.TowerConnector - Tower GET 'https://api.cloud.stage-seqera.io/user-info' - can refresh=true; msgId=37101dcb74cf; tokens=JwtAuth{key='fa657055338a584c9483fd732bb0da01', endpoint='https://api.cloud.stage-seqera.io', bearer='eyJ0aWQiOiAyNTV9LmYyNjBjO..', refresh='null', createdAt=null, updatedAt=null}
2024-12-16 10:56:20.829	09:56:20.829 [virtual-executor115] TRACE i.s.w.t.c.connector.TowerConnector - Tower GET 'https://api.cloud.stage-seqera.io/user-info' - can refresh=true; msgId=d827eea40d17; tokens=JwtAuth{key='fa657055338a584c9483fd732bb0da01', endpoint='https://api.cloud.stage-seqera.io', bearer='eyJ0aWQiOiAyNTV9LmYyNjBjO..', refresh='null', createdAt=null, updatedAt=null}

```

File: src/main/groovy/io/seqera/wave/service/CredentialServiceImpl.groovy [hunk 1]
```diff
@@ -92,7 +92,7 @@ class CredentialServiceImpl implements CredentialsService {
         // log for debugging purposes
         log.debug "Credentials matching criteria registryName=$registryName; userId=$identity.userId; workspaceId=$identity.workspaceId; endpoint=$identity.towerEndpoint => $creds"
```
Comment (munishchouhan) [inline src/main/groovy/io/seqera/wave/service/CredentialServiceImpl.groovy:93]: should we log workflowId too?

Comment (pditommaso) [reply src/main/groovy/io/seqera/wave/service/CredentialServiceImpl.groovy:93]: The logic of this change is to use two caches, one short (when workflowId is absent) and the long (when workflowId) is provided.

The cache key should determined by the `cacheKey` parameter which is computed ad-hoc c   

https://github.com/seqeralabs/wave/blob/9122a302cf63d11fb086916b4d59fe239385f363/src/main/groovy/io/seqera/wave/tower/client/TowerClient.groovy#L52-L60

In the case of user-info is made by the URI and the auth key 

https://github.com/seqeralabs/wave/blob/9122a302cf63d11fb086916b4d59fe239385f363/src/main/groovy/io/seqera/wave/tower/client/TowerClient.groovy#L64-L64

In principle the `cacheKey` should be identical and therefore there should be only one request. Either the cacheKey is not unique or micronaut cache is failing somewhere. 

I'd suggest making a local test and checking with multiple concurrent requests

Comment (munishchouhan) [reply src/main/groovy/io/seqera/wave/service/CredentialServiceImpl.groovy:93]: ok sure
I will work on it

Comment (munishchouhan) [reply src/main/groovy/io/seqera/wave/service/CredentialServiceImpl.groovy:93]: I tried adding custom keyGenerator to make sure that its only using the cacheKey as key, but still not resolved
https://github.com/micronaut-projects/micronaut-cache/issues/71#issuecomment-583538488

digging mre into it

File: VERSION [hunk 2]
```diff
@@ -1 +1 @@
-1.15.5
+1.15.6-B5
```
Comment (pditommaso) [inline VERSION:1]:
```suggestion
1.15.5
```

File: src/main/groovy/io/seqera/wave/encoder/MoshiExchange.groovy [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package io.seqera.wave.encoder
+
+/**
+ * Marker interface for Moshi encoder objects
```
Comment (pditommaso) [inline src/main/groovy/io/seqera/wave/encoder/MoshiExchange.groovy]:
```suggestion
 * Marker interface for Moshi encoded exchange objects
```

File: src/main/groovy/io/seqera/wave/store/cache/AbstractTieredCache.groovy [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    private AsyncCache<String,V> l1
+
+    private final Duration ttl
+
+    private L2TieredCache<String,String> l2
+
+    private final Lock sync = new ReentrantLock()
+
+    AbstractTieredCache(L2TieredCache<String,String> l2, MoshiEncodeStrategy encoder, Duration duration, long maxSize) {
+        log.info "Cache configuring '${getName()}' - prefix=${getPrefix()}; ttl=${duration}; max-size: ${maxSize}; l2=${l2}"
```
Comment (pditommaso) [inline src/main/groovy/io/seqera/wave/store/cache/AbstractTieredCache.groovy:65]:
```suggestion
        log.info "Cache '${getName()}' config - prefix=${getPrefix()}; ttl=${duration}; max-size: ${maxSize}; l2=${l2}"
```

**Pull request merged**